### PR TITLE
feat: optimistic concurrency control loop for IAM

### LINKED ISSUE #613: Consider a wrapper for optimistic concurrency control.
The `SetIamPolicy()` functions use OCC ([Optimistic Concurrency Control](https://en.wikipedia.org/wiki/Optimistic_concurrency_control)) to change the IAM policies.

Typically the application is expected to write a loop like this:

```
while (!given_up()) {
  auto current = client.GetIamPolicy(...);
  if (!current) continue; // try again, but we can discuss that...
  // ... make some changes to "current" ...
  auto updated = client.SetIamPolicy(..., *std::move(current));
  if (updated) return updated; // yay!
  if (!updated for OCC failure) continue; // try again
  if (!updated for some permanent error) return updated.status();
}
```

Feels like we could wrap that in a function, something like:

```
// Run the OCC loop calling mutator each time, returning an unset optional indicates that
// no changes are needed.
StatusOr<Policy> ChangeIamPolicy(std::function<optional<Policy>(Policy)> mutator);
```

### PR DESCRIPTION
**Pull request by coryan**
Adds functions to DatabaseAdminClient and InstanceAdminClient that run a
optimistic concurrency control (OCC) loop for IAM updates. That makes
the customer code a bit easier to write and less error prone. This is
analogous to the Commit() loops for the spanner::Client.

This fixes #613 

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp-spanner/1162)

### CODE REVIEW

File: google/cloud/spanner/database_admin_client.cc [hunk 1]
```diff
@@ -64,6 +64,57 @@ StatusOr<google::iam::v1::Policy> DatabaseAdminClient::SetIamPolicy(
...
+        return current_policy;
+      }
+      desired->set_etag(std::move(etag));
+      auto result = SetIamPolicy(db, *std::move(desired));
+      if (RerunnablePolicy::IsOk(result.status())) {
+        return result;
+      }
+      last_status = std::move(result).status();
+    }
+    if (!rerun_policy->OnFailure(last_status)) {
```
Comment (devbww) [inline google/cloud/spanner/database_admin_client.cc]: Nit: It is a little funny that `!OnFailure(s)` means `IsPermanentFailure(s) || IsExhausted()`, so, if we continue, we know we saw `!IsExhausted()`.  Yet, the very next thing we test is `!IsExhausted()`.  For a "limited count" policy this is redundant.  For a "limited time" policy it means we're immediately sampling the clock again, which feels redundant/racy.

We could address that, but the bigger question might be whether any rerun policy should be able to trigger an "Exhausted before first call".  After all, it isn't really a "rerun" if it was never a "run".

File: google/cloud/spanner/instance_admin_client.cc [hunk 2]
```diff
@@ -67,6 +67,57 @@ StatusOr<google::iam::v1::Policy> InstanceAdminClient::SetIamPolicy(
...
+        return current_policy;
+      }
+      desired->set_etag(std::move(etag));
+      auto result = SetIamPolicy(in, *std::move(desired));
+      if (RerunnablePolicy::IsOk(result.status())) {
+        return result;
+      }
+      last_status = std::move(result).status();
+    }
+    if (!rerun_policy->OnFailure(last_status)) {
```
Comment (devbww) [inline google/cloud/spanner/instance_admin_client.cc]: Ditto.

File: google/cloud/spanner/samples/samples.cc [hunk 3]
```diff
@@ -266,42 +266,42 @@ void AddDatabaseReader(google::cloud::spanner::InstanceAdminClient client,
...
+
+        auto member_pos = std::find(binding.members().begin(),
+                                    binding.members().end(), new_reader);
+        if (member_pos != binding.members().end()) {
+          std::cout << "The entity " << new_reader
+                    << " is already a database reader:\n"
+                    << current.DebugString();
+          return {};
+        }
+        if (member_pos == binding.members().end()) {
```
Comment (devbww) [inline google/cloud/spanner/samples/samples.cc]: Isn't this `if (true)`?

Comment (coryan) [reply google/cloud/spanner/samples/samples.cc]: Fixed.

File: google/cloud/spanner/database_admin_client.cc (hunk 1 above)
Comment (coryan) [reply google/cloud/spanner/database_admin_client.cc]: I think it is funny too.  WDTY of the do/while loop instead?

File: google/cloud/spanner/instance_admin_client.cc (hunk 2 above)
Comment (coryan) [reply google/cloud/spanner/instance_admin_client.cc]: Changed this too.

File: google/cloud/spanner/database_admin_client.h [hunk 4]
```diff
@@ -244,6 +242,50 @@ class DatabaseAdminClient {
   StatusOr<google::iam::v1::Policy> SetIamPolicy(
       Database db, google::iam::v1::Policy policy);
 
+  /**
+   * Updates the IAM policy for an instance using an optimistic concurrency
+   * control loop.
+   *
+   * This function repeatedly reads the current IAM policy in @p in, and then
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/spanner/database_admin_client.h]: Fixed.

File: google/cloud/spanner/CMakeLists.txt [hunk 5]
```diff
@@ -206,7 +206,8 @@ add_library(
     value.h
     version.cc
     version.h
-    version_info.h)
+    version_info.h
+    iam_updater.h)
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/spanner/CMakeLists.txt]: Fixed.

[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/spanner/CMakeLists.txt]: And we should have cmake-format do this for us, I will check on that later.

File: google/cloud/spanner/database_admin_client.cc (hunk 1 above)
Comment (devbww) [reply google/cloud/spanner/database_admin_client.cc]: I like it better.  But things like ...
```
  do {
   ...
    if (!rerun_policy->OnFailure(last_status)) break;
    std::this_thread::sleep_for(backoff_policy->OnCompletion());
  } while (!rerun_policy->IsExhausted());
```
reinforce the idea that the rerun and backoff policies should be unified.  It is no use sleeping beyond the deadline, for example.

The main feature would be that `rerun_policy->OnFailure(status)` would either return `false` _immediately_ on a permanent error, or `false` _when_ the count/deadline has been exceeded, or `true` _when_ the next attempt should be made.

But that is something for another day.

File: google/cloud/spanner/instance_admin_client.cc [hunk 6]
```diff
@@ -93,8 +93,7 @@ StatusOr<google::iam::v1::Policy> InstanceAdminClient::SetIamPolicy(
 
   Status last_status =
       Status(StatusCode::kUnknown, "Exhausted before first call");
```
Comment (devbww) [inline google/cloud/spanner/instance_admin_client.cc]: It looks like this doesn't need such specific initialization now.

**Pull request merged**